titus-ext/zookeeper/src/main/java/com/netflix/titus/ext/zookeeper/supervisor/ZookeeperLeaderElector.java

```diff
@@ -106,8 +106,17 @@ public boolean join() {
 
     @Override
     public boolean leaveIfNotLeader() {
-        LeaderElectionProcess process = leaderElectionProcessRef.getAndSet(null);
-        return process != null && process.leaveIfNotLeader();
+        synchronized (leaderElectionProcessRef) {
```

Comment:
Do we need this change? The previous version seemed good to me. Worst case when `join()` races with `leaveIfNotLeader()` can be treated as if `leaveIfNotLeader()` was called before the instance joined the leader election process.

Reply:
The previous version had a bug, since it did `getAndSet(null)` even if the node was a leader. When I was fixing that bug, I chose more defensive implementation with synchronized.

Reply:
Discussed in person. This is handling the case when the instance is already the leader, and we don't want to clean up the `leaderElectionProcessRef` in that case.